pandas/tests/groupby/test_groupby_nan.py

```diff
@@ -0,0 +1,413 @@
+        result = df.groupby(['a', 'b'], dropna=False).sum()
+
+        # expected result
+        data = [[12.30, 233., 12.], [13., 13., 13.], [123.23, 123., 1.]]
+        index = MultiIndex(levels=[['a', 'b'], [np.nan, 'a', 'b']],
+                           labels=[[0, 0, 1], [0, 2, 1]],
+                           names=['a', 'b'])
+        expected = DataFrame(data, index=index, columns=['c', 'd', 'e'])
+
+        assert_frame_equal(result, expected)
+
+    def test_with_na(self):
+        """
+        Compare behavior with and without the dropna=False option
+        """
+
+        index = Index(np.arange(10))
+
+        for dtype in ['float64', 'float32', 'int64', 'int32', 'int16', 'int8',
+                      'datetime64[ns]']:
+            values = Series(np.ones(10), index, dtype=dtype)
+            labels = Series([nan, 'foo', 'bar', 'bar', nan, nan, 'bar',
+                             'bar', nan, 'foo'], index=index)
+
+            # this SHOULD be an int
+            grouped = values.groupby(labels)
+            agged = grouped.agg(len)
+            expected = Series([4, 2], index=['bar', 'foo'])
+
+            assert_series_equal(agged, expected, check_dtype=False)
+
+            # self.assertTrue(issubclass(agged.dtype.type, np.integer))
+
+            # explicity return a float from my function
+            def f(x):
+                return float(len(x))
+
+            agged = grouped.agg(f)
+            expected = Series([4, 2], index=['bar', 'foo'])
+
+            assert_series_equal(agged, expected, check_dtype=False)
+            # self.assertTrue(issubclass(agged.dtype.type, np.float))
+
+            # GH 3729: keeping the NaNs
+
+            # this SHOULD be an int
+            grouped = values.groupby(labels, dropna=False)
+            agged = grouped.agg(len)
+            expected = Series([4, 4, 2], index=[nan, 'bar', 'foo'])
+
+            assert_series_equal(agged, expected, check_dtype=False)
+            # self.assertTrue(issubclass(agged.dtype.type, np.integer))
+
+            # explicity return a float from my function
+            def f(x):
+                return float(len(x))
+
+            agged = grouped.agg(f)
+            expected = Series([4, 4, 2], index=[nan, 'bar', 'foo'])
+
+            assert_series_equal(agged, expected, check_dtype=False)
+            # self.assertTrue(issubclass(agged.dtype.type, np.float)
+
+    def test_groupby_transform_with_nan_group(self):
+        # GH 9941
+        df = pd.DataFrame({'a': range(10),
+                           'b': [1, 1, 2, 3, np.nan, 4, 4, 5, 5, 5]})
+        result = df.groupby(df.b)['a'].transform(max)
+        expected = pd.Series([1., 1., 2., 3., np.nan, 6., 6., 9., 9., 9.],
+                             name='a')
+        assert_series_equal(result, expected)
+
+        result = df.groupby(df.b, dropna=False)['a'].transform(max)
+        expected = pd.Series([1, 1, 2, 3, 4, 6, 6, 9, 9, 9],
+                             name='a', dtype=np.integer)
```

Comment:
@jreback is this (ensuring `dtype=np.integer`) an admissible way to deal with windows's int32/int64 issues? Without that, the test fail because result is int32 and expected is int64... which is maybe something I should fix, instead of tampering with the tests, but I have no idea how to do that. In this pull/request, I never actually touch the data...